File: chapter1/007/main.py
def main():
    matrix = [
        [1,1,1,1],
        [2,2,2,2],
        [3,3,3,3],
        [4,4,4,4]
    ]
    print_matrix(matrix)

    matrix = [
        [1,1,1],
        [2,2,2],
        [3,3,3]
    ]
    print_matrix(matrix)

    matrix = [
        [1,1],
        [2,2]
    ]
    print_matrix(matrix)

    matrix = [
        [1]
    ]
    print_matrix(matrix)

def rotate(matrix):
    middle = int(len(matrix) / 2)
    for i in range(middle):
        rotate_matrix(matrix, i, len(matrix) - i - 1)

def rotate_matrix(matrix, start, end):
    for i in range(end - start):
        # top with left exchange
        matrix[start][start+i], matrix[end-i][start] = matrix[end-i][start], matrix[start][start+i]

        # right with top exchange
        matrix[start+i][end], matrix[start][start+i] = matrix[start][start+i], matrix[start+i][end]

        # botton with right exchange
        matrix[end][end-i], matrix[start+i][end] = matrix[start+i][end], matrix[end][end-i]

def print_matrix(matrix):
    for l in matrix:
        print(l)
    print("--------------------")
    rotate(matrix)
    for l in matrix:
        print(l)
    print("===================+")

main()

File: chapter1/007/main.go
package main

import (
	"fmt"
	"math"
	"sync"
)

func main() {
	matrix := [][]int32{
		{1, 2, 3, 0, 0, 0},
		{4, 5, 6, 0, 0, 0},
		{7, 8, 9, 0, 0, 0},
		{10, 11, 12, 0, 0, 0},
		{10, 11, 12, 0, 0, 0},
		{10, 11, 12, 0, 0, 0},
	}
	printMatrix(matrix)

	matrix = [][]int32{
		{1, 2, 3, 0},
		{4, 5, 6, 0},
		{7, 8, 9, 0},
		{10, 11, 12, 0},
	}
	printMatrix(matrix)

	matrix = [][]int32{
		{1, 2, 3},
		{4, 5, 6},
		{7, 8, 9},
	}
	printMatrix(matrix)

	matrix = [][]int32{
		{1, 2},
		{3, 4},
	}
	printMatrix(matrix)

	matrix = [][]int32{
		{1},
	}
	printMatrix(matrix)
}

func rotate(matrix *[][]int32) {
	middle := int(math.Round(float64(len(*matrix)/2.0)) - 1)
	var wg sync.WaitGroup
	wg.Add(middle + 1)
	for i := 0; i <= middle; i++ {
		go func(i int) {
			rotateStartEnd(matrix, i, len(*matrix)-1-i)
			wg.Done()
		}(i)
	}
	wg.Wait()
}

func rotateStartEnd(matrix *[][]int32, start, end int) {
	m := *matrix
	for i := 0; i < end-start; i++ {
		m[start][start+i], m[end-i][start] = m[end-i][start], m[start][start+i]
	}
	for i := 0; i < end-start; i++ {
		m[start+i][end], m[start][start+i] = m[start][start+i], m[start+i][end]
	}
	for i := 0; i < end-start; i++ {
		m[end][end-i], m[start+i][end] = m[start+i][end], m[end][end-i]
	}
}

func printMatrix(matrix [][]int32) {
	for _, l := range matrix {
		for _, e := range l {
			fmt.Printf("%d\t", e)
		}
		fmt.Println()
	}
	rotate(&matrix)
	fmt.Println("---------------------------------")
	for _, l := range matrix {
		for _, e := range l {
			fmt.Printf("%d\t", e)
		}
		fmt.Println()
	}
	fmt.Println("===================================")
}
